### PR TITLE
[KeyVault] Managed HSM public preview

### PR DESCRIPTION
**Pull request by bim-msft**
HSM: Hardware Security Module.

### Module Changes:
- **Core**
   1. cloud @arrownj @fengzhou-msft @haroldrandom @jiasli @jsntcy @Juliehzl @MyronFanQiu 
   2. client_factory @arrownj @fengzhou-msft @haroldrandom @jiasli @jsntcy @Juliehzl @MyronFanQiu 
   3. _shared @arrownj @fengzhou-msft @haroldrandom @jiasli @jsntcy @Juliehzl @MyronFanQiu
- **KeyVault Module**
   1. @fengzhou-msft @yungezz 
- **Live Test**
   1. acr @fengzhou-msft @yungezz 
   2. appconfig
   3. appservice @Juliehzl @qwordy 
   4. cdn
   5. cosmosdb
   6. network @haroldrandom @MyronFanQiu 
   7. resource @Juliehzl @zhoxing-ms 
   8. storage @jsntcy @Juliehzl @zhoxing-ms 
   9. vm @arrownj @qwordy 

### New features:
- **[HSM management plane operations](https://microsoft.sharepoint.com/:w:/r/teams/AzureKeyVault-SDKPSHCLICollaboration/_layouts/15/Doc.aspx?sourcedoc=%7Bf547b138-df5b-40ae-97b5-333f75152a9d%7D&action=edit&wdPid=2617f046&cid=39dfe167-f153-4a99-8aae-0298aac1e2fd)**
    1. Create: `az keyvault create --hsm-name {hsm_name}`
    2. Delete: `az keyvault delete --hsm-name {hsm_name}`
    3. Show: `az keyvault show --hsm-name {hsm_name}`
    4. Update: `az keyvault update-hsm --hsm-name {hsm_name}`
    5. List
         1. All resources: `az keyvault list`
         2. HSM only: `az keyvault list --resource-type hsm`
         3. Vault only: `az keyvault list --resource-type vault`
- **HSM operations for keys**
    1. Create
         1. `az keyvault key create --hsm-name {hsm_name} -n {key_name}`
         2. `az keyvault key create --id {hsm_base_url}/keys/{key_name}`
    2. Delete
         1. `az keyvault key delete --hsm-name {hsm_name} -n {key_name}`
         2. `az keyvault key delete --id {hsm_base_url}/keys/{key_name}`
    3. Show
         1. `az keyvault key show --hsm-name {hsm_name} -n {key_name}`
         2. `az keyvault key show --id {hsm_base_url}/keys/{key_name}`
    4. List
         1. `az keyvault key list --hsm-name {hsm_name}`
         2. `az keyvault key list --id {hsm_base_url}`
    5. Backup/restore/recover/list-deleted/purge/set-attributes/import/encrypt/decrypt...
- **[HSM full backup and restore](https://microsoft.sharepoint.com/:w:/t/AzureKeyVault-SDKPSHCLICollaboration/EfmaAEXo0I1Pl5kJRwwo3zoBS5koQWdXyeF5Vo1AHStlSA?e=FStdkK)**
    1. Backup
         1. `az keyvault backup start --hsm-name/--id {hsm_name}/{hsm_base_url} --storage-account-name {} --blob-container-name {} --storage-container-SAS-token {}` 
    2. Restore
          1. `az keyvault restore start --hsm-name/--id {hsm_name}/{hsm_base_url} --storage-account-name {} --blob-container-name {} --storage-container-SAS-token {} --backup-folder {}` 
- **[HSM security domain (disaster recovery)](https://microsoft.sharepoint.com/:w:/t/AzureKeyVault-SDKPSHCLICollaboration/EfmaAEXo0I1Pl5kJRwwo3zoBS5koQWdXyeF5Vo1AHStlSA?e=FStdkK)**
    1. Download the security domain file from the old HSM
         1. `az keyvault security-domain download --hsm-name/--id {hsm_name}/{hsm_base_url} --sd-wrapping-keys {cert1} {cert2} {cert3} ... --sd-quorum {} --security-domain-file {}`
    2. Start recovery mode on a new HSM
         1. `az keyvault security-domain init-recovery --hsm-name/--id {hsm_name}/{hsm_base_url} --sd-exchange-key {}`
    3. Start the recovery
         1. `az keyvault security-domain upload --hsm-name/--id {hsm_name}/{hsm_base_url} --sd-file {} --sd-exchange-key {} --sd-wrapping-keys {private_key1} {private_key2} ... --passwords {pwd_for_key1} {pwd_for_key2} ...`
- **[HSM access control (RBAC mode 2)](https://microsoft.sharepoint.com/:w:/t/AzureKeyVault-SDKPSHCLICollaboration/EV5K6lixtDxIqagvSWJplkMBXPOLkd9l7Iv0oHlXEbC57A)**
    1. List role definitions
         1. `az keyvault role definition list --hsm-name/--id {hsm_name}/{hsm_base_url}`
    2. Create role assignments
         1. `az keyvault role assignment create --hsm-name/--id {hsm_name}/{hsm_base_url} --role {} --assignee {} --scope {}`
    3. Delete role assignments
         1. `az keyvault role assignment delete --hsm-name/--id {hsm_name}/{hsm_base_url} --role {} --assignee {} --scope {}`
    4. List role assignments
         1. `az keyvault role assignment list --hsm-name/--id {hsm_name}/{hsm_base_url} --role {} --assignee {} --scope {}`

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yungezz** [comment]
move to S168 to align with service schedule.

**Comment by yungezz** [comment]
based on latest info in service team, public preview date maybe june-july not locked down yet,  GA is even later. move this one to backlog since CLI work done already.

File: Dockerfile [hunk 1]
```diff
@@ -1,7 +1,7 @@
-#---------------------------------------------------------------------------------------------
+# --------------------------------------------------------------------------------------------
```
Comment (yungezz) [inline Dockerfile:1]: why this change

File: src/azure-cli/azure/cli/command_modules/keyvault/_client_factory.py [hunk 2]
```diff
@@ -3,23 +3,104 @@
...
+    else:
+        class_name = 'KeyVaultClient.'
+    return KEYVAULT_TEMPLATE_STRINGS[resource_type].format(
+        api_version=api_version,
+        module_name=module_name,
+        class_name=class_name,
+        obj_name='{}')
+
+
+def get_client_factory(resource_type, client_name=''):
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_client_factory.py:68]: I feel maybe this part of code could be abstracted to a common class by _client_factory_vendor.py for future local development work.

File: src/azure-cli/azure/cli/command_modules/keyvault/_help.py [hunk 3]
```diff
@@ -12,6 +12,16 @@
 short-summary: Manage KeyVault keys, secrets, and certificates.
 """
 
+helps['keyvault backup'] = """
+type: group
+short-summary: Manage full HSM backup
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_help.py:17]: shouldn't it be keyvault hsm backup?

File: src/azure-cli/azure/cli/command_modules/keyvault/_help.py [hunk 4]
```diff
@@ -390,14 +423,24 @@
...
+short-summary: Update the properties of a Vault.
 examples:
-  - name: Update the properties of a key vault. (autogenerated)
+  - name: Update the properties of a Vault. (autogenerated)
     text: |
         az keyvault update --enabled-for-disk-encryption true --name MyKeyVault --resource-group MyResourceGroup
     crafted: true
 """
 
+helps['keyvault update-hsm'] = """
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_help.py:434]: shouldn't it be hsm backup? I think it's necessary to have a hsm sub group

File: src/azure-cli/azure/cli/command_modules/keyvault/_help.py [hunk 5]
```diff
@@ -406,3 +449,53 @@
     text: |
         az keyvault wait --name MyVault --created
 """
+
+helps['keyvault wait-hsm'] = """
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_help.py:453]: hsm create --wait?

File: src/azure-cli/azure/cli/command_modules/keyvault/_help.py [hunk 6]
```diff
@@ -406,3 +449,53 @@
...
+type: command
+short-summary: Place the CLI in a waiting state until a condition of the HSM is met.
+examples:
+  - name: Pause CLI until the HSM is created.
+    text: |
+        az keyvault wait-hsm --name MyHSM --created
+"""
+
+helps['keyvault security-domain'] = """
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_help.py:462]: any abbreviation for security-domain?

File: src/azure-cli/azure/cli/command_modules/keyvault/_params.py [hunk 7]
```diff
@@ -54,22 +53,57 @@ class CLIJsonWebKeyOperation(str, Enum):
...
     # ARGUMENT DEFINITIONS
     vault_name_type = CLIArgumentType(
-        help='Name of the key vault.', options_list=['--vault-name'], metavar='NAME', id_part=None,
+        help='Name of the Vault.', options_list=['--vault-name'], metavar='NAME', id_part=None,
         completer=get_resource_name_completion_list('Microsoft.KeyVault/vaults'))
 
+    deleted_vault_name_type = CLIArgumentType(
+        help='Name of the deleted Vault.', options_list=['--vault-name'], metavar='NAME', id_part=None)
+
+    hsm_name_type = CLIArgumentType(help='Name of the HSM.', type=get_hsm_base_url_type(self.cli_ctx),
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_params.py:83]: why using hsm instead of mhsm in cli command interface?

File: src/azure-cli/azure/cli/command_modules/keyvault/_params.py [hunk 8]
```diff
@@ -255,78 +391,185 @@ class CLIJsonWebKeyOperation(str, Enum):
...
+            c.extra('identifier', options_list=['--id'], validator=validate_vault_or_hsm, help='Id of the HSM.')
+
+    with self.argument_context('keyvault security-domain init-recovery') as c:
+        c.argument('sd_exchange_key', help='Local file path to store the exported key.')
+
+    with self.argument_context('keyvault security-domain upload') as c:
+        c.argument('sd_file', help='This file contains security domain encrypted using SD Exchange file downloaded '
+                                   'in security-domain init-recovery command.')
+        c.argument('sd_exchange_key', help='The exchange key for security domain.')
+        c.argument('sd_wrapping_keys', nargs='*',
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_params.py:443]: file validator?

File: src/azure-cli/azure/cli/command_modules/keyvault/_params.py [hunk 9]
```diff
@@ -255,78 +391,185 @@ class CLIJsonWebKeyOperation(str, Enum):
...
+                   help='Space-separated file paths to PEM files containing private keys.')
+        c.argument('passwords', nargs='*', help='Space-separated password list for --sd-wrapping-keys. '
+                                                'CLI will match them in order. Can be omitted if your keys are without '
+                                                'password protection.')
+
+    with self.argument_context('keyvault security-domain download') as c:
+        c.argument('sd_wrapping_keys', nargs='*',
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_params.py:450]: file validator

File: src/azure-cli/azure/cli/command_modules/keyvault/_params.py [hunk 10]
```diff
@@ -255,78 +391,185 @@ class CLIJsonWebKeyOperation(str, Enum):
...
+    with self.argument_context('keyvault security-domain wait') as c:
+        c.argument('hsm_base_url', hsm_name_type, help='Name of the HSM. Can be omitted if --id is specified.',
+                   required=False)
+        c.argument('identifier', options_list=['--id'], validator=validate_vault_or_hsm, help='Id of the HSM.')
+        c.argument('resource_group_name', options_list=['--resource-group', '-g'],
+                   help='Proceed only if HSM belongs to the specified resource group.')
+    # endregion
+
+    # region keyvault backup/restore
+    for item in ['backup', 'restore']:
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_params.py:466]: if this backup/restore for hsm only? if yes, why not hsm backup restore

File: src/azure-cli/azure/cli/command_modules/keyvault/_params.py [hunk 11]
```diff
@@ -255,78 +391,185 @@ class CLIJsonWebKeyOperation(str, Enum):
...
+                c.argument('vault_base_url', hsm_name_type, required=False,
+                           help='Name of the HSM. Can be omitted if --id is specified.')
+                c.extra('identifier', options_list=['--id'], validator=validate_vault_or_hsm, help='Id of the HSM.')
+                c.ignore('cls')
+
+    with self.argument_context('keyvault backup start', arg_group='Storage Id') as c:
+        c.argument('storage_resource_uri', required=False, validator=process_storage_uri,
+                   help='Azure Blob storage container Uri. If specified all other \'Storage Id\' arguments '
+                        'should be omitted')
+        c.extra('storage_account_name', help='Name of Azure Storage Account.')
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_params.py:478]: with storage blob container uri, storage accoutn name and blob container name could be parsed from it. so either storage_resource_uri or (storage_accoutn_name+blob_container_name) right?

File: src/azure-cli/azure/cli/command_modules/keyvault/_params.py [hunk 12]
```diff
@@ -399,7 +648,46 @@ class CLIJsonWebKeyOperation(str, Enum):
...
+        with self.argument_context('keyvault certificate {}'.format(item)) as c:
             c.argument('maxresults', options_list=['--maxresults'], type=int)
+
+    with self.argument_context('keyvault role') as c:
+        c.argument('scope',
+                   help='scope at which the role assignment or definition applies to, '
+                        'e.g., "/" or "/keys" or "/keys/{keyname}"')
+
+    with self.argument_context('keyvault role', arg_group='Id') as c:
+        c.argument('hsm_base_url', hsm_name_type)
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_params.py:661]: I think Amit mentioned for hsm, use hsm_name only

File: src/azure-cli/azure/cli/command_modules/keyvault/_validators.py [hunk 13]
```diff
@@ -29,21 +29,34 @@ def _extract_version(item_id):
...
     """
     from azure.cli.core.profiles import ResourceType
     from msrestazure.tools import parse_resource_id
 
-    client = get_mgmt_service_client(cli_ctx, ResourceType.MGMT_KEYVAULT).vaults
-    for vault in client.list():
-        id_comps = parse_resource_id(vault.id)
-        if 'name' in id_comps and id_comps['name'].lower() == vault_name.lower():
-            return id_comps['resource_group']
+    if vault_name:
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_validators.py:43]: validate vault_name and hsm_name is mutual exclusive?

File: src/azure-cli/azure/cli/command_modules/keyvault/custom.py [hunk 14]
```diff
@@ -190,15 +205,83 @@ def _scaffold_certificate_profile(cmd):
     return template
 
 
-def list_keyvault(client, resource_group_name=None):
+def delete_vault_or_hsm(cmd, client, resource_group_name=None, vault_name=None, hsm_name=None):
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/custom.py:208]: for each func below, I think it's necessary to check valut_name and hsm_name is mutual exclusive

File: src/azure-cli/azure/cli/command_modules/keyvault/custom.py [hunk 15]
```diff
@@ -277,7 +360,46 @@ def get_default_policy(cmd, client, scaffold=False):  # pylint: disable=unused-a
...
+    if hsm_name:
+        hsm_client = get_client_factory(ResourceType.MGMT_PRIVATE_KEYVAULT, Clients.managed_hsms)(cmd.cli_ctx, None)
+        return recover_hsm(cmd=cmd,
+                           client=hsm_client,
+                           resource_group_name=resource_group_name,
+                           location=location,
+                           hsm_name=hsm_name)
+
+
+def recover_hsm(cmd, client, hsm_name, resource_group_name, location):
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/custom.py:381]: what's difference with recover_hsm and recover_vault_or_hsm with hsm_name?

File: Dockerfile (hunk 1 above)
Comment (bim-msft) [reply Dockerfile:1]: Weird. Maybe something wrong with my IDE. Will revert this.

File: src/azure-cli/azure/cli/command_modules/keyvault/_validators.py [hunk 16]
```diff
@@ -115,6 +130,45 @@ def process_secret_set_namespace(cmd, namespace):
     namespace.value = content
 
 
+def process_storage_uri(ns):
+    if not ns.storage_resource_uri:
+        if ns.storage_account_name and ns.blob_container_name:
+            ns.storage_resource_uri = 'https://{}.blob.core.windows.net/{}'.format(
```
Comment (arrownj) [inline src/azure-cli/azure/cli/command_modules/keyvault/_validators.py:136]: `core.windows.net` is storage endpoint for public cloud. Are you sure it's a fixed value here ?

File: src/azure-cli/azure/cli/command_modules/keyvault/security_domain/jwe.py [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+from cryptography.hazmat.primitives import hashes, padding
+from cryptography.hazmat.primitives.asymmetric import padding as asymmetric_padding
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+from knack.util import CLIError
+
+from .sp800_108 import KDF
+from .utils import Utils
+
+
+class JWEHeader:  # pylint: disable=too-many-instance-attributes
```
Comment (arrownj) [inline src/azure-cli/azure/cli/command_modules/keyvault/security_domain/jwe.py:22]: Let's define new class as a normal Python3 class (extend from object).

File: src/azure-cli/azure/cli/command_modules/keyvault/security_domain/jwe.py [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+        hash_data = bytearray()
+        hash_data.extend(header_bytes)
+        hash_data.extend(self.jwe_decode.init_vector)
+        hash_data.extend(self.jwe_decode.ciphertext)
+        hash_data.extend(KDF.to_big_endian_64bits(auth_bits))
+
+        hMAC = hmac.HMAC(hk, msg=hash_data, digestmod=hashlib.sha512)
+        return hMAC.digest()
+
+    def Aes256HmacSha512Encrypt(self, cek, plaintext):
```
Comment (arrownj) [inline src/azure-cli/azure/cli/command_modules/keyvault/security_domain/jwe.py:193]: Any special reason to define this non-pythonic function name ?

File: src/azure-cli/setup.py [hunk 19]
```diff
@@ -8,6 +8,7 @@
 from __future__ import print_function
 from codecs import open
 from setuptools import setup, find_packages
+import os
```
Comment (arrownj) [inline src/azure-cli/setup.py:11]: unused import ?

File: src/azure-cli/azure/cli/command_modules/keyvault/security_domain/jwe.py (hunk 17 above)
Comment (arrownj) [reply src/azure-cli/azure/cli/command_modules/keyvault/security_domain/jwe.py:22]: OK, we can skip reviewing these code first.

File: src/azure-cli/setup.py [hunk 20]
```diff
@@ -138,7 +140,9 @@
     'websocket-client~=0.56.0',
     'xmltodict~=0.12',
     'javaproperties==0.5.1',
-    'jsondiff==1.2.0'
+    'jsondiff==1.2.0',
+    'azure-mgmt-core~=1.2.0',
```
Comment (arrownj) [inline src/azure-cli/setup.py:144]: `azure-mgmt-core` and `azure-core` are already in azure-cli-core dependencies, we can remove them here.

File: src/azure-cli-core/azure/cli/core/commands/client_factory.py [hunk 21]
```diff
@@ -22,6 +22,13 @@ def resolve_client_arg_name(operation, kwargs):
         logger.info("Keyword 'client_arg_name' is deprecated and should be removed.")
         return kwargs['client_arg_name']
     path, op_path = operation.split('#', 1)
+
+    # Support vendored sdks
+    # eg: azure.cli.command_modules.keyvault.vendored_sdks
+    items = path.split('.')
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/commands/client_factory.py:28]: We can reuse `path_comps`.

```py
    if path_comps[0] == 'azure':
        if (path_comps[1] != 'cli' or
                len(path_comps) >= 5 and path_comps[4] == 'vendored_sdks'):
            # Public SDK: azure.mgmt.resource... (mgmt-plane) or azure.storage.blob... (data-plane)
            # Vendored SDK: azure.cli.command_modules.keyvault.vendored_sdks...
            client_arg_name = 'self'
        else:
            # CLI custom method: azure.cli.command_modules.resource...
            client_arg_name = 'client'
```

Comment (bim-msft) [reply src/azure-cli-core/azure/cli/core/commands/client_factory.py:28]: Graceful code!

File: src/azure-cli/setup.py (hunk 20 above)
Comment (bim-msft) [reply src/azure-cli/setup.py]: Removed

File: src/azure-cli/setup.py (hunk 19 above)
Comment (bim-msft) [reply src/azure-cli/setup.py]: Good catch, removed!

**Review by jiasli — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py [hunk 22]
```diff
@@ -247,28 +337,285 @@ def test_keyvault_mgmt(self, resource_group):
                  checks=[self.check('properties.enableSoftDelete', True),
                          self.check('properties.enablePurgeProtection', True)])
 
-        # test '--enable-rbac-authorization'
-        # temporarily disable this since our test subscription doesn't support RBAC authorization
-        """
+
+class KeyVaultHSMSecurityDomainScenarioTest(ScenarioTest):
+    @unittest.skip('Hard to make it idempotent to run recording/live.')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py:342]: prefer @record_only() if currently resource is hard to create

File: src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py [hunk 23]
```diff
@@ -247,28 +337,285 @@ def test_keyvault_mgmt(self, resource_group):
...
+        })
+        self.cmd('az storage account create -n {storage_account} -g {rg}')
+        self.cmd('az storage container create -n {blob} --account-name {storage_account} -g {rg}')
+
+        self.kwargs['sas'] = '?' + self.cmd('az storage account generate-sas --start {sas_start} --expiry {sas_expiry} '
+                                            '--https-only '
+                                            '--permissions rwdlacu --resource-types sco --services b '
+                                            '--account-name {storage_account}').get_output_in_json().replace('%3A', ':')
+
+        self.cmd('az keyvault backup start --id {hsm_url} --blob-container-name {blob} '
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py:423]: it seems get_output_in_json() is unused.

File: src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py [hunk 24]
```diff
@@ -247,28 +337,285 @@ def test_keyvault_mgmt(self, resource_group):
...
+            'hsm_name': ACTIVE_HSM_NAME,
+            'storage_account': self.create_random_name('clitesthsmsa', 24),
+            'blob': self.create_random_name('clitesthsmblob', 24),
+            'sas_start': (datetime.utcnow() - timedelta(minutes=30)).strftime('%Y-%m-%dT%H:%M:%SZ'),
+            'sas_expiry': (datetime.utcnow() + timedelta(minutes=30)).strftime('%Y-%m-%dT%H:%M:%SZ')
+        })
+        self.cmd('az storage account create -n {storage_account} -g {rg}')
+        self.cmd('az storage container create -n {blob} --account-name {storage_account} -g {rg}')
+
+        self.kwargs['sas'] = '?' + self.cmd('az storage account generate-sas --start {sas_start} --expiry {sas_expiry} '
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py:418]: May I know if encoded sas works for keyvault scenario?

File: src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py [hunk 25]
```diff
@@ -247,28 +337,285 @@ def test_keyvault_mgmt(self, resource_group):
...
+                                            '--https-only '
+                                            '--permissions rwdlacu --resource-types sco --services b '
+                                            '--account-name {storage_account}').get_output_in_json().replace('%3A', ':')
+
+        self.cmd('az keyvault backup start --id {hsm_url} --blob-container-name {blob} '
+                 '--storage-account-name {storage_account} '
+                 '--storage-container-SAS-token "{sas}"',
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py:425]: typically sas related command is hard to playback because there is random signature. Does it work in your scenrio?

File: src/azure-cli/azure/cli/command_modules/keyvault/_help.py [hunk 26]
```diff
@@ -143,34 +163,171 @@
...
+"""
+
 helps['keyvault key'] = """
 type: group
 short-summary: Manage keys.
 """
 
+helps['keyvault key delete'] = """
+type: command
+short-summary: Deletes a key of any type from storage in Vault or HSM.
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/keyvault/_help.py:191]: Shall we use the first person voice, right?
```suggestion
short-summary: Delete a key of any type from storage in Vault or HSM.
```

File: src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py (hunk 25 above)
Comment (bim-msft) [reply src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py:425]: @Juliehzl This works. I generate a new SAS every time.

File: src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py (hunk 24 above)
Comment (bim-msft) [reply src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py:418]: @Juliehzl What does "encoded" mean? Without `.replace('%3A', ':')`? It didn't work, so I decoded it.

File: src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py (hunk 23 above)
Comment (bim-msft) [reply src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py:423]: @Juliehzl Good idea.

File: src/azure-cli/azure/cli/command_modules/keyvault/_params.py [hunk 27]
```diff
@@ -399,7 +647,46 @@ class CLIJsonWebKeyOperation(str, Enum):
...
+        c.argument('role_assignment_name', options_list=['--name', '-n'], help='Name of the role assignment.')
+        c.argument('assignee', help='represent a user, group, or service principal. '
+                                    'supported format: object id, user sign-in name, or service principal name')
+        c.argument('assignee_object_id',
+                   help='Use this parameter instead of \'--assignee\' to bypass graph permission issues. '
+                        'This parameter only works with object ids for users, groups, service principals, and '
+                        'managed identities. For managed identities use the principal id. For service principals, '
+                        'use the object id and not the app id.')
+        c.argument('ids', nargs='+', help='space-separated role assignment ids')
+        c.argument('role', help='role name or id')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/keyvault/_params.py:675]: Just curious, may I ask why Key Vault has the separate RBAC command?
Why not use `az role xxx`?

Comment (bim-msft) [reply src/azure-cli/azure/cli/command_modules/keyvault/_params.py:675]: @zhoxing-ms Good question. I'm not very clear about the exact motivation of the service team. One possible reason is that they didn't want to expose HSM related access controls to the outer world (AAD), so they implemented some similar kinds of stuff which can under their own control.

**Review by arrownj — APPROVED** [review]
Looking forward to version 2 👍

File: src/azure-cli/azure/cli/command_modules/keyvault/_help.py [hunk 28]
```diff
@@ -143,34 +163,171 @@
...
+long-summary: Retrieves a list of the keys in the Vault or HSM as JSON Web Key structures that contain the
+    public part of a deleted key. This operation includes deletion-specific information. The Get Deleted Keys
+    operation is applicable for vaults enabled for soft-delete. While the operation can be invoked on any
+    Vault or HSM, it will return an error if invoked on a non soft-delete enabled Vault or HSM. This operation
+    requires the keys/list permission.
+"""
+
+helps['keyvault key purge'] = """
+type: command
+short-summary: Permanently deletes the specified key.
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/keyvault/_help.py:278]: some short description missing `Vault or HSM`.

File: src/azure-cli-core/azure/cli/core/commands/client_factory.py [hunk 29]
```diff
@@ -22,12 +22,16 @@ def resolve_client_arg_name(operation, kwargs):
...
         return kwargs['client_arg_name']
     path, op_path = operation.split('#', 1)
+
     path_comps = path.split('.')
     if path_comps[0] == 'azure':
-        # for CLI command modules
-        # SDK method: azure.mgmt.foo... or azure.foo...
-        # custom method: azure.cli.command_modules.foo...
-        client_arg_name = 'client' if path_comps[1] == 'cli' else 'self'
+        if path_comps[1] != 'cli' or len(path_comps) >= 5 and path_comps[4] == 'vendored_sdks':
```
[the remark this answers is not in the document]
Comment (bim-msft) [reply src/azure-cli-core/azure/cli/core/commands/client_factory.py:28]: @yungezz Maybe a helper function would be better, will add this.

File: src/azure-cli/azure/cli/command_modules/keyvault/_help.py (hunk 28 above)
Comment (bim-msft) [reply src/azure-cli/azure/cli/command_modules/keyvault/_help.py:278]: @yungezz The original text doesn't contain `vault` or `key vault`, so I didn't modify the description.

File: src/azure-cli-core/azure/cli/core/commands/client_factory.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (bim-msft) [reply src/azure-cli-core/azure/cli/core/commands/client_factory.py:28]: @yungezz Added a helper function.

```python
def _is_vendored_sdk_path(path_comps):
    return len(path_comps) >= 5 and path_comps[4] == 'vendored_sdks'


def resolve_client_arg_name(operation, kwargs):
    ...
    if path_comps[0] == 'azure':
        if path_comps[1] != 'cli' or _is_vendored_sdk_path(path_comps):
            # Public SDK: azure.mgmt.resource... (mgmt-plane) or azure.storage.blob... (data-plane)
            # Vendored SDK: azure.cli.command_modules.keyvault.vendored_sdks...
            client_arg_name = 'self'
            ...
```

File: src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py (hunk 24 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/keyvault/tests/latest/test_keyvault_commands.py:419]: I see.

**Review by Juliehzl — APPROVED** [review]
LGTM for code owner part

**Review by yungezz — APPROVED** [review]

**Review by yungezz — APPROVED** [review]

**Review by haroldrandom — APPROVED** [review]
Network LGTM

**Review by fengzhou-msft — APPROVED** [review]
